Feature/kubernetes clients (#620)
ordina-jworks/ordina-jworks.github.io: _posts/2020-08-06-kubernetes-clients-comparison.md

Screenshot: Details about a pod in Lens, including Prometheus supplied metrics
<img alt="Lens pod details" src="{{ '/img/2020-08-06-kubernetes-clients-comparison/lens-pod-details.png' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width: 1000px;">

Screenshot: Overview of a cluster in Lens, including the last error events
<img alt="Lens cluster overview" src="{{ '/img/2020-08-06-kubernetes-clients-comparison/lens-cluster-overview.png' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width: 1000px;">

## Kubenav

## Infra App
[Infra App](https://infra.app/){:target="_blank" rel="noopener noreferrer"} is a new addition to the list of Kubernetes clients.
It is made by the people over at Docker Desktop & [Kitematic](https://kitematic.com){:target="_blank" rel="noopener noreferrer"} and is being developed behind closed doors, which has been addressed as unpleasant within the Kubernetes community.
It provides you with a clean, simplistic user interface which groups everything you need to know about a single resource together.

> Everything speaks for itself and all the information you need is within the reach of a few simple clicks.

Author: Everything is self explanatory and all the information you need is available within a few simple clicks

> When you open the application for the first time, they greet you by asking you to fill in your e-mail address.

Author: you are greeted with a prompt asking you for your e-mail address

> This is probably for newsletters and updates, but I wish this step was optional.

Author: Although, this is.... , I wish

You quickly notice that there is only basic functionality available in the application, which makes sense as the client is still in early access at the time of writing.
You can browse resources per namespace, go through application logs, read and edit YAML configuration and check the current resources used by your deployment.
There is a metrics interface for the whole cluster aswell, which supplies you with a structured and detailed view about your nodes.

> The application is still very young and right now, it is lacking some functionality that you might expect or find in other clients.

Author: Since the application is still very, it's lacking...

This view provides you with an easy overview of the resources within the cluster and (super useful) provides a list of the last seen error events in the cluster.

Navigating to the list of pods provides an overview of all pods in the cluster.
By clicking on a pod you're provided with the details of that pod (`kubectl describe`).
From here, you can directly dive into the pod logs, shell into the pod, make edits or remove the pod from the cluster.

Similar support is available for most common resources within the cluster: Statefulsets, deployment, configmaps, secrets, ...
The workflow is always the following: open the type in the sidebar on the left, click on an object to get details.
From that detailed view, certain actions can be performed on the object.

As with most of the tools in this comparison, Lens is quite feature-rich.
The most common types are support and common actions are available.
But Lens has another trick up its sleeve which makes it different from the other tools: Metrics/Prometheus integration.

The integration relies on a Prometheus instance being installed in the cluster that exposes the supported metrics.

> You can opt for Lens to install Prometheus (and other required components) for you, but in real scenarios, you either don't have those rights or you'll already have a Prometheus installed in the cluster.

Reviewer: ```suggestion
You can opt for Lens to install Prometheus (and other required components) for you, but in real scenarios, you either don't have those rights or you'll already have a Prometheus instance installed in the cluster.
```

In those cases, you can just [add some configuration](https://github.com/lensapp/lens/blob/master/troubleshooting/custom-prometheus.md){:target="_blank" rel="noopener noreferrer"} to that instance and point the Lens app to that Prometheus instance.

> It will use port-forwarding under the hood, so no need to expose the Prometheus instance to the outside.

Reviewer: ```suggestion
It will use port-forwarding under the hood, so no need to expose the Prometheus instance to the outside world.
```

The integration works nicely and instantly provides some metrics about your cluster and deployed components.
This provides good insights for developers to figure out their resource consumption without leaving their Kubernetes client.
The charts and data seem to be very rudimentary, but improvement are expected to arrive over time.

> As most of the client described in this post, Lens app is an open-source project.

Reviewer 2: clients*

Mirantis is behind the development of Lens, but at the time of writing, no supported (paid) version is available.

> There is continuous active development on the app on the app and releases are about one month apart, so bugs fixes and new features should be available regularly.

Reviewer: ```suggestion
There is continuous active development on the app and releases are about one month apart, so bug fixes and new features should be available regularly.
```

> There is very active development happing on the app and releases are about one month apart, so bugs and new features should be available regularly.

Reviewer 2: happening*

Reviewer 2: "so bugs and new features should be available regularly". die bugs ertussen vind ik raar, of ligt dat aan mij? :D

They share the same codebase, so the navigation should be similar on all platforms.

The navigation menu is self-explanatory as it categorizes every Kubernetes resource, which you can then filter by namespace in the top-right corner.
They support all widely used Kubernetes resources and the status and configuration of those resources. 
You can easily switch between Kubernetes clusters by using the drop-down menu available in the menu.

There is a window that is available at all times where you can consult different deployment logs and a terminal. Each window is categorized under a tab.
When you minimize this window, you will notice a blue shell icon in the right bottom corner of Kubenav, which allows you to open your current logs or terminal again.

The app is made in the Ionic Framework using Capacitator as cross-platform runtime, which is something you notice when you start using the app.
I don't really like this feeling, as I prefer my clients to have a more native feeling, although I understand why the developers chose this approach.

> The primary factor of why I would use this application is because they support almost every resource combined with a GUI if you don't like using the terminal.

Reviewer: ```suggestion
The primary reason why I would use this application is because they support almost every resource combined with a GUI if you don't like using the terminal.
```

However, I'm not a fan of the interface in general, as it seems to be messy and overwhelming at times when a resource provides you with a great deal of information in an uncategorized way.
If you want to know every detail about your resource at all times, then this application is definitely for you. I didn't have a reason yet to use the mobile client, but I'm sure that depends on your use case.

<img alt="Kubenav slideshow" src="{{ '/img/2020-08-06-kubernetes-clients-comparison/kubenav.gif' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width: 1000px;">
<p style="text-align: center; font-style: italic;">Kubenav slideshow</p>

## Infra App
[Infra App](https://infra.app/){:target="_blank" rel="noopener noreferrer"} is a new addition to the list of Kubernetes clients.

> It is made by the people over at Docker Desktop & [Kitematic](https://kitematic.com){:target="_blank" rel="noopener noreferrer"} and is being developed behind closed doors, which has been addressed as unpleasant within the Kubernetes community.

Reviewer: ```suggestion
It is made by the people over at Docker Desktop & [Kitematic](https://kitematic.com){:target="_blank" rel="noopener noreferrer"} and is being developed behind closed doors, which has been addressed as "unpleasant" within the Kubernetes community.
```

It provides you with a clean, simplistic user interface that groups everything you need to know about a single resource together.
Everything is self-explanatory and all the information you need is available within a few simple clicks.

When you open the application for the first time, you are greeted with a prompt asking you for your e-mail address.
Although this is probably for newsletters and updates, I wish this step was optional.

> You quickly notice that there is only basic functionality available in the application, which makes sense as the client is still in early access at the time of writing.

Reviewer: ```suggestion
You quickly notice that only basic functionality is available in the application, which makes sense as the client is still in early access at the time of writing.
```

You can browse resources per namespace, go through application logs, read and edit YAML configuration, and check the current resources used by your deployment.
There is a metrics interface for the whole cluster as well, which supplies you with a structured and detailed view about your nodes.

> Since the application is still very young right now, it is lacking some functionality that you might expect or find in other clients.

Reviewer: ```suggestion
Since the application is still very young, it is lacking some functionality that you might expect or find in other clients.
```

If you want something with more than basic functionality right now, this might not be the application you are looking for.
However, if it has what you need, you will find that it will be very easy and straightforward to manage your Kubernetes cluster with this client.

<img alt="Kubenav slideshow" src="{{ '/img/2020-08-06-kubernetes-clients-comparison/infra-app.gif' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width: 1000px;">
<p style="text-align: center; font-style: italic;">Infra slideshow</p>

## Conclusion

Now the real question, which client should you use.
As with any question about software, it depends.
If you like to be lightning-fast and don't mind struggling through a steep learning curve, K9s might be a tool for you.

> It's the personal favorite of the authors this post, meanly because of it's shortcuts and lightning-fast load times.

Reviewer: ```suggestion
It's the personal favorite of the authors this post, mainly because of it's shortcuts and lightning-fast load times.
```

If you are using some software that has plugins available for Octant, definitely give it a try. One of the authors prefers Octant as non-terminal GUI.
The [plugins](https://github.com/topics/octant-plugin){:target="_blank" rel="noopener noreferrer"} add a lot of value to the tool and might make it a very compelling option for your use-case.

If you often need to optimize your resource usage, want a client that just works, and is easy to use, go for Lens.

> This is definitely hitting the sweet spot between ease of use, stability, and available feature set.

Reviewer: ```suggestion
This definitely hits the sweet spot between ease of use, stability, and available feature set.
```
